### ISSUE TITLE
Exception: Specified a Platform for a Context which does not support all required kernels

### ISSUE THREAD

**Issue by ca-taylor** [maccallumlab/meld#78]
I built meld using openmm 7.4.2 with CUDA 11.1.   When I try to run on an nvidia DGX A100 platform, I get the failure mode below.  I admittedly do not know much about GPU support under OpenMM and thought someone here could point me in the right direction.  

Note that a similar build using CUDA 10.0.130 works well on our 2080ti GPUs.  

Is it telling me that the A100 is an unsupported platform or is there something else going on that I can remedy?  

/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/simtk/openmm/app/internal/amber_file_parser.py:1018: UserWarning: Non-optimal GB parameters detected for GB model GBn2
  warnings.warn(
Traceback (most recent call last):
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/bin/launch_remd", line 29, in <module>
    main()
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/bin/launch_remd", line 25, in main
    launch.launch(console, args.debug, args.console_log)
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/Meld-0.4.14-py3.8.egg/meld/remd/launch.py", line 108, in launch
    remd_runner.run(communicator, system_runner, store)
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/Meld-0.4.14-py3.8.egg/meld/remd/leader.py", line 116, in run
    system_runner.prepare_for_timestep(0., self._step)
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/Meld-0.4.14-py3.8.egg/meld/system/openmm_runner/runner.py", line 112, in prepare_for_timestep
    self._initialize_simulation()
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/Meld-0.4.14-py3.8.egg/meld/system/openmm_runner/runner.py", line 243, in _initialize_simulation
    self._simulation = _create_openmm_simulation(
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/Meld-0.4.14-py3.8.egg/meld/system/openmm_runner/runner.py", line 384, in _create_openmm_simulation
    return Simulation(topology, system, integrator, platform, properties)
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/simtk/openmm/app/simulation.py", line 105, in __init__
    self.context = mm.Context(self.system, self.integrator, platform, platformProperties)
  File "/apps/cuda/11.1.0/intel/2020.0.166/openmpi/4.0.4/meld/0.4.14/lib64/python3.8/site-packages/simtk/openmm/openmm.py", line 15677, in __init__
    this = _openmm.new_Context(*args)
Exception: Specified a Platform for a Context which does not support all required kernels

**Comment by jlmaccal** [comment]
I don't have any A100's so we haven't encountered this, but there is no reason that A100s can't be supported?

How was this built / installed? I suspect the issue is that the `meldplugin` has not been built for cuda11, which is probably required for A100.

**Comment by frank-stonybrook** [comment]
Could you run `python -m simtk.testInstallation` on a GPU compute node? What is the output and how many platforms it report results on?

**Comment by jlmaccal** [comment]
Since I haven't heard anything further, I'm going to close this.

**Comment by ca-taylor** [comment]
I'm sorry, this got dropped for a bit but it is still an issue with my latest build (0.4.19) on top of OpenMM 7.5.0.     All the OpenMM tests pass and...

`[chasman@c44a-s1 test]$ python -m simtk.testInstallation

OpenMM Version: 7.5
Git Revision: Unknown

There are 3 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 1.96957e-06
Reference vs. CUDA: 6.73439e-06
CPU vs. CUDA: 6.21782e-06

All differences are within tolerance.
`

**Comment by ca-taylor** [comment]
I can't figure out why meld continues to report that there is the required kernels are not supported.

**Comment by ca-taylor** [comment]
> I can't figure out why meld continues to report that there is the required kernels are not supported.

I think I'm missing something very trivial.  I just don't know what it is?    How does meld/openmm decide that a given platform (CUDA) does not support "all required kernels"?

**Comment by ca-taylor** [comment]
-------------------------------------------------------------------------------
[chasman@c44a-s1 test]$ 
[chasman@c44a-s1 test]$ python
Python 3.8.6 (default, Sep 29 2020, 12:37:04) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from simtk.openmm import *
>>> 
>>> print(pluginLoadedLibNames)
('/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCPU.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMRPMDCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMDrudeCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMAmoebaCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCudaCompiler.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMRPMDReference.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMDrudeReference.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMAmoebaReference.so')
>>> 
>>> print(Platform.getPluginLoadFailures())
()
>>>--------------------------------------------------------------------------------------------------------------

In other words, everything looks correct to me.   Hmmmm?

**Comment by frank-stonybrook** [comment]
Did you local environment has CUDA devices? How did you set up the **OPENMM_CUDA_COMPILER** environment?

**Comment by ca-taylor** [comment]
> Did you local environment has CUDA devices? 

No, the machine I build on does not have GPUs.  It has a full CUDA environment with a stubbed libcuda.so  but no GPUs.   

> How did you set up the **OPENMM_CUDA_COMPILER** environment?

This is what I do when building in the meld plugin directory.  

export OPENMM_DIR=${HPC_OPENMM_DIR}
export OPENMM_INCLUDE_PATH=${HPC_OPENMM_INC}
export OPENMM_LIB_PATH=${HPC_OPENMM_LIB}

#export LD_LIBRARY_PATH=${HPC_OPENMM_LIB}:$LD_LIBRARY_PATH
export OPENMM_CUDA_COMPILER=nvcc
export EIGEN3_INCLUDE_DIR=/apps/eigen3/3.3.4/include

rm -rf build
mkdir build
cd build

#  -DCMAKE_INSTALL_PREFIX="${HOME}/apps/meld" \
#  -DCMAKE_INSTALL_PREFIX=${HPC_OPENMM_DIR} \

cmake .. \
  -DCMAKE_INSTALL_PREFIX=/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/meld/0.4.19 \
  -DCMAKE_VERBOSE_MAKEFILE=ON \
  -DOPENMM_DIR=${HPC_OPENMM_DIR} \
  -DEXTRA_COMPILE_FLAGS="-L${HPC_CUDA_LIB}/stubs -lcuda" \
  -DSWIG_EXECUTABLE=${HPC_SWIG_BIN}/swig
make
make install
make PythonInstall

**Comment by frank-stonybrook** [comment]
Does this testing local environment ﻿**chasman@c44a-s1** have GPUs?

**Comment by frank-stonybrook** [comment]
> I'm sorry, this got dropped for a bit but it is still an issue with my latest build (0.4.19) on top of OpenMM 7.5.0. All the OpenMM tests pass and...
> 
> `[chasman@c44a-s1 test]$ python -m simtk.testInstallation
> 
> OpenMM Version: 7.5
> Git Revision: Unknown
> 
> There are 3 Platforms available:
> 
> 1 Reference - Successfully computed forces
> 2 CPU - Successfully computed forces
> 3 CUDA - Successfully computed forces
> 
> Median difference in forces between platforms:
> 
> Reference vs. CPU: 1.96957e-06
> Reference vs. CUDA: 6.73439e-06
> CPU vs. CUDA: 6.21782e-06
> 
> All differences are within tolerance.
> `

Based on this testing results, it seems that your installation is correct. 

When does MELD throw "the required kernels are not supported" error? Usually it means OpenMM can not find a valid CUDA device.

**Comment by ca-taylor** [comment]
> When does MELD throw "the required kernels are not supported" error? Usually it means OpenMM can not find a valid  CUDA device.

Below is my remd.log output which should give you an idea of when the exception is thrown.   

`c45a-s21.ufhpc:001 2021-04-30 14:38:44 INFO meld.remd.launch: Launching replica exchange on follower
c45a-s21.ufhpc:001 2021-04-30 14:38:44 INFO meld.remd.launch: Meld version is 0.4.19
c45a-s21.ufhpc:001 2021-04-30 14:38:44 INFO meld.remd.launch: OpenMM_Meld version is 7.5.0.dev-Unknown
c45a-s21.ufhpc:001 2021-04-30 14:38:44 INFO meld.remd.launch: Loading system
c45a-s21.ufhpc:000 2021-04-30 14:38:44 INFO meld.remd.launch: Launching replica exchange on leader
c45a-s21.ufhpc:000 2021-04-30 14:38:44 INFO meld.remd.launch: Meld version is 0.4.19
c45a-s21.ufhpc:000 2021-04-30 14:38:44 INFO meld.remd.launch: OpenMM_Meld version is 7.5.0.dev-Unknown
c45a-s21.ufhpc:000 2021-04-30 14:38:44 INFO meld.remd.launch: Loading system
c45a-s21.ufhpc:001 2021-04-30 14:38:44 INFO meld.remd.launch: Loading run options
c45a-s21.ufhpc:000 2021-04-30 14:38:44 INFO meld.remd.launch: Loading run options
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.comm: c45a-s21.ufhpc found cuda devices: 2,3
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.comm: CUDA_VISIBLE_DEVICES is set.
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.comm: hostname: c45a-s21.ufhpc, device_id: 0
c45a-s21.ufhpc:000 2021-04-30 14:38:45 DEBUG meld.comm: negotiate_device_id took 24.476 ms
c45a-s21.ufhpc:001 2021-04-30 14:38:45 INFO meld.comm: c45a-s21.ufhpc found cuda devices: 2,3
c45a-s21.ufhpc:001 2021-04-30 14:38:45 INFO meld.comm: hostname: c45a-s21.ufhpc, device_id: 1
c45a-s21.ufhpc:001 2021-04-30 14:38:45 DEBUG meld.comm: negotiate_device_id took 24.544 ms
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.remd.leader: Beginning replica exchange
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.remd.leader: Running replica exchange step 1 of 1000.
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.system.openmm_runner.runner: Creating implicit solvent system
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.system.openmm_runner.runner: Using a cutoff of 1.8
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.system.openmm_runner.runner: Enabling hydrogen mass=4, constraining all bonds
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.system.openmm_runner.runner: Using "gbNeck2" implicit solvent
c45a-s21.ufhpc:000 2021-04-30 14:38:45 INFO meld.system.openmm_runner.runner: Creating integrator with 4.5 fs timestep
c45a-s21.ufhpc:000 2021-04-30 14:38:45 CRITICAL meld.comm: MPI node 1/2 raised exception.
`

**Comment by frank-stonybrook** [comment]
I didn't see any connection between the output you provided and required kernel not support issue you mentioned here. This output is very vague.

**Comment by ca-taylor** [comment]
> I didn't see any connection between the output you provided and required kernel not support issue you mentioned here. This output is very vague.
> 
> Everything until the last line looks total normal. What is the directly evidence of the kernel issue you mentioned?

Just the last line where the exception is raised.   I'm sorry if it wasn't helpful.  I'm not very well-versed in running MELD and am just trying to help the Perez group get MELD working on the UF HiperGator computing system.   

I don't see anything in the build system that tells what CUDA compute capabilities to target.  I'm used to seeing that specifiied somewhere when building "FAT" binaries but I suppose for JIT it is not required if that is what the MELD/OpenMM plugins rely on.

**Comment by jlmaccal** [comment]
I'm still not totally sure what is going on, but just to clear up one thing: meld is not reporting `Exception: Specified a Platform for a Context which does not support all required kernels`. That error is coming from OpenMM and occurs when you try to run a calculation on a platform where OpenMM can't find all of the required kernels.

A few more details that might help in understanding what is happening... OpenMM has a modular architecture. It support multiple platforms, including `CPU`, `Reference`, `CUDA`, etc. Obviously, `CUDA` is the most relevant here. Further, OpenMM supports "plugins" that extend its functionality. MELD has such a plugin, creatively called `meldplugin` that implements the MELD forces in OpenMM. The problem appears to be that OpenMM is unable to find the MELD plugin for the CUDA environment. I'm not yet sure what the cause is.

The `python -m simtk.testInstallation` script is checking to see if it can find implementations of the standard OpenMM forces, but isn't checking for the MELD plugin at all.

However, this does *not* look correct:
```
('/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCPU.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMRPMDCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMDrudeCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMAmoebaCUDA.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMCudaCompiler.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMRPMDReference.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMDrudeReference.so',
'/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/openmm/7.5.0/lib64/plugins/libOpenMMAmoebaReference.so')
```

There should be an entry for `libMeldPlugin` and `libMeldPluginCUDA`.

Actually, now that I'm looking at things, this does not look correct:
```
-DCMAKE_INSTALL_PREFIX=/apps/cuda/11.0.207/intel/2020.0.166/openmpi/4.0.4/meld/0.4.19 
```

This is going to install the MELD plugin into that directory, but it won't be part of OpenMM's search path and so the plugin will not be found. Typically, we install the two meld plugin libraries to the openmm plugin dir. Off-hand, I'm not sure if it is possible to specify an alternate search location.

**Comment by jlmaccal** [comment]
@peastman Can you comment on if there is anyway for the OpenMM plugin dir to be specified (e.g. as an environment variable). Is it possible to search multiple directories?

**Comment by ca-taylor** [comment]
> There should be an entry for libMeldPlugin and libMeldPluginCUDA.

Apparently that was the  missing piece.   I moved the MELD plugins to the OpenMM plugin directory and that got me past the "Platform Context" error.  Of course, now it dumps core after calling,
  
                    OpenMM::CudaArray::upload<float4>()  ???:0

but that is a different problem which I can work on.  

I've learned something useful about OpenMM + MELD so, thank you!

**Comment by peastman** [comment]
> Can you comment on if there is anyway for the OpenMM plugin dir to be specified (e.g. as an environment variable).

`OPENMM_PLUGIN_DIR`.

> Is it possible to search multiple directories?

You can call `Platform.loadPluginsFromDirectory()` multiple times for multiple directories.  The Python module calls it automatically for the default location, but that's just a convenience.   Take a look at

https://github.com/openmm/openmm/blob/d8ef57fed6554ec95684e53768188e1f666405c9/wrappers/python/openmm/__init__.py#L24-L27

**Comment by ca-taylor** [comment]
We've got it working now so feel free to close this one.   Thank you for all the replies and insight.
